core/arch/instrlist.c

```diff
@@ -481,15 +492,18 @@ instrlist_encode_to_copy(dcontext_t *dcontext, instrlist_t *ilist, byte *copy_pc
     instr_t *inst;
     int len = 0;
 #ifdef ARM
-    /* XXX i#1734: reset encdoe state to avoid any stale encode state
+    /* XXX i#1734: reset encode state to avoid any stale encode state
      * or dangling pointer.
      */
     if (instr_get_isa_mode(instrlist_first(ilist)) == DR_ISA_ARM_THUMB)
         encode_reset_it_block(dcontext);
 #endif
-    if (has_instr_jmp_targets || max_pc != NULL) {
+    if (IF_DEBUG_ELSE(true, has_instr_jmp_targets) || max_pc != NULL) {
         /* must set note fields first with offset, or compute length */
         for (inst = instrlist_first(ilist); inst; inst = instr_get_next(inst)) {
+            if (IF_DEBUG_ELSE(!has_instr_jmp_targets, false)
+                    && TEST(INSTR_OPERANDS_VALID, (inst)->flags))
+                DOCHECK(3, { check_inst_opnds(inst); });
```

Comment:
This causes the instrlist walk at -checklevel 0 though: we want the walk to be a higher checklevel.

I suggest just making a completely separate walk inside "DOCHECK(2" (voting for 2 b/c it's the default, so people will see the assert, unless they deliberately set a different -checklevel, in which case they'll get the speed they're looking for).